### PR TITLE
Add cors proxy plugin 🏄

### PR DESCRIPTION
**Pull request by oponder**
This lets you start up Happa and have it talk to a production api and/or passage endpoint with just a one liner:

```
$ HAPPA_PROXY_INSTALLATION=ginger yarn start
```

```
yarn run v1.21.1
$ ./node_modules/.bin/webpack-dev-server --progress --colors --config ./webpack.dev.js
🏄  [Proxy Plugin] Checking for HAPPA_PROXY_INSTALLATION or HAPPA_PROXY_BASE_DOMAIN
🏄  [Proxy Plugin: api] Starting CORS proxy on localhost:8000 to: https://api.g8s.ginger.eu-west-1.aws.gigantic.io:443
🏄  [Proxy Plugin: api] Succesfully started CORS proxy
🏄  [Proxy Plugin: passage] Starting CORS proxy on localhost:5001 to: https://passage.g8s.ginger.eu-west-1.aws.gigantic.io:443
🏄  [Proxy Plugin: passage] Succesfully started CORS proxy

... rest of the build ...

🏄  [Proxy Plugin: api] Request Proxied -> /v4/user/
🏄  [Proxy Plugin: api] Request Proxied -> /v4/info/
🏄  [Proxy Plugin: api] Request Proxied -> /v4/organizations/
...
🏄  [Proxy Plugin: passage] Request Proxied -> /recovery/
.. rest of the requests and rebuilds ...
```

It runs a cors proxy which means you do not need to do anything tricky with your browser safety settings.

### CODE REVIEW

**Comment by oponder** [comment]
Yeah technically we sometimes do. Not often. But I’ll tidy it up a bit and add passage as well.

Will probably change the env variable naming since passage doesn’t need an audience setting, and I want to ya e the proxy related env vars names similarly

File: scripts/api-proxy.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+  // remove trailing slash
+  var cleanProxyUrl = proxyUrl.replace(/\/$/, '');
+  // remove all forward slashes
+  var cleanProxyPartial = proxyPartial.replace(/\//g, '');
+
+  proxy.use('/' + cleanProxyPartial, function (req, res) {
+    try {
+      console.log(
+        chalk.green('🏄🏽‍♂️  [API Proxy Plugin]'),
```
Comment (ubergesundheit) [inline scripts/api-proxy.js:21]: Btw this is how it looks on my machine:
image

Maybe we should change the surfer to "Person surfing" 🏄 without skin tone for compatibility?

```suggestion
        chalk.green('🏄  [API Proxy Plugin]'),
```

File: webpack.dev.js [hunk 2]
```diff
@@ -21,7 +31,51 @@ module.exports = merge(common, {
...
-  plugins: [new webpack.HotModuleReplacementPlugin()],
+  plugins: [
+    new webpack.HotModuleReplacementPlugin(),
+    {
+      apply: (compiler) => {
+        compiler.hooks.afterEnvironment.tap(
+          'Start proxy if HAPPA_AUDIENCE is defined',
+          () => {
+            console.log(
+              chalk.green('🏄🏽‍♂️  [API Proxy Plugin]'),
```
Comment (ubergesundheit) [inline webpack.dev.js:42]:
```suggestion
              chalk.green('🏄  [API Proxy Plugin]'),
```

File: webpack.dev.js [hunk 3]
```diff
@@ -21,7 +31,51 @@ module.exports = merge(common, {
...
+            const envFileVars = dotenv.config().parsed;
+            const { HAPPA_AUDIENCE } = Object.assign(
+              {},
+              envFileVars,
+              process.env
+            );
+
+            if (!HAPPA_AUDIENCE) {
+              console.log(
+                chalk.green('🏄🏽‍♂️  [API Proxy Plugin]'),
```
Comment (ubergesundheit) [inline webpack.dev.js:55]:
```suggestion
                chalk.green('🏄  [API Proxy Plugin]'),
```

File: webpack.dev.js [hunk 4]
```diff
@@ -21,7 +31,51 @@ module.exports = merge(common, {
...
+                'Skipping. HAPPA_AUDIENCE not defined. '
+              );
+
+              return;
+            }
+
+            console.log(
+              chalk.green('🏄🏽‍♂️  [API Proxy Plugin]'),
```
Comment (ubergesundheit) [inline webpack.dev.js:63]:
```suggestion
              chalk.green('🏄  [API Proxy Plugin]'),
```

File: webpack.dev.js [hunk 5]
```diff
@@ -21,7 +31,51 @@ module.exports = merge(common, {
...
+              'Starting CORS proxy on localhost:8000 to:',
+              HAPPA_AUDIENCE
+            );
+
+            apiProxy.startProxy(8000, HAPPA_AUDIENCE, '', false, '*');
+
+            console.log(
+              chalk.green('🏄🏽‍♂️  [API Proxy Plugin]'),
```
Comment (ubergesundheit) [inline webpack.dev.js:71]:
```suggestion
              chalk.green('🏄  [API Proxy Plugin]'),
```

**Comment by ubergesundheit** [comment]
Here is another idea, maybe you've already thought of this:

It would be awesome if it would be possible to override certain responses with this for example to mock error states or malformed responses.

**Comment by oponder** [comment]
That's a cool idea! but I'll save that maybe for another PR

**Comment by oponder** [comment]
Alright! applied your suggestions for the surfer and made it possible to set up a 2nd proxy for passage.

The env vars are now: HAPPA_API_PROXY and HAPPA_PASSAGE_PROXY.

Updated the first comment.

**Comment by ubergesundheit** [comment]
Sorry for dumping another idea here, but I'm thinking this could also be used to have azure, kvm & aws instances running at the same time.

**Comment by oponder** [comment]
Ok it is now: 

`HAPPA_PROXY_INSTALLATION=ginger yarn start`

I also made an alternative: 

`HAPPA_PROXY_BASE_DOMAIN=g8s.ginger.eu-west-1.aws.gigantic.io yarn start`

Because I also wanted it to be really easy for people that want to contribute to happa but don't have opsctl set up to get started. (like some helpful customers for example)

Updated the first comment again.

It's a bit more complicated than I wanted it to be, mostly because I wanted to avoid calling `opsctl` multiple times, and there are two plugins that need to make use of the information that `opsctl` returns.

File: webpack.common.js [hunk 6]
```diff
@@ -6,24 +6,34 @@ const HtmlWebpackPlugin = require('html-webpack-plugin');
...
 const dotenv = require('dotenv');
+const getServiceURL = require('./scripts/webpack/getServiceURL');
 
 const envFileVars = dotenv.config().parsed;
 
 const makeEndpoints = () => {
   const defaults = {
     HAPPA_API_ENDPOINT: 'http://localhost:8000',
     HAPPA_PASSAGE_ENDPOINT: 'http://localhost:5001',
+    PROXY: undefined,
```
Comment (ubergesundheit) [inline webpack.common.js:17]: Is this a leftover?
```suggestion
    HAPPA_PROXY_INSTALLATION: undefined,
    HAPPA_PROXY_BASE_DOMAIN: undefined,
```

Comment (oponder) [reply webpack.common.js:17]: Yeah, doing a quick little refactoring pass

**Pull request merged**